pkg/runs/runsDownload.go

```diff
@@ -0,0 +1,188 @@
+			err = galasaErrors.NewGalasaError(galasaErrors.GALASA_ERROR_CANNOT_OVERWRITE_FILE, targetFilePath)
+			return err
+		}
+
+		// Set up the directory structure and artifact file on the host's file system
+		var newFile io.Writer = nil
+		newFile, err = CreateEmptyArtifactFile(fileSystem, targetFilePath)
+		if err == nil {
+			log.Printf("Writing artifact '%s' to '%s' on local file system", fileName, targetFilePath)
+
+			// Set up a byte buffer to gradually read the downloaded file to avoid out-of-memory issues.
+			reader := bufio.NewReader(fileDownloaded)
+			buffer := make([]byte, bufferCapacity)
+
+			// Read a chunk of the downloaded file into the buffer and write the buffer's contents to the
+			// newly-created file.
+			for {
+				bytesRead, err := reader.Read(buffer)
+				if err != nil {
+					if err == io.EOF {
+						// There was nothing else to read.
+						log.Printf("Artifact '%s' written to '%s' OK", fileName, targetFilePath)
+						break
+					}
+					return err
+				}
+
+				_, err = newFile.Write(buffer[:bytesRead])
+				if err != nil {
+					err = galasaErrors.NewGalasaError(galasaErrors.GALASA_ERROR_FAILED_TO_WRITE_FILE, targetFilePath, err.Error())
+					return err
+				}
+			}
+		}
+	}
+	return err
+}
+
+// Creates an empty file representing an artifact that is being written. Any parent directories that do not exist
+// will be created. Returns the empty file that was created or an error if any file creation operations failed. 
+func CreateEmptyArtifactFile(fileSystem utils.FileSystem, targetFilePath string) (io.Writer, error) {
+	
+	var err error = nil
+	var newFile io.Writer = nil
+
+	targetDirectoryPath := filepath.Dir(targetFilePath)
+	err = fileSystem.MkdirAll(targetDirectoryPath)
+	if err != nil {
+		err = galasaErrors.NewGalasaError(galasaErrors.GALASA_ERROR_FAILED_TO_CREATE_FOLDERS, targetDirectoryPath, err.Error())
+	} else {
+		newFile, err = fileSystem.Create(targetFilePath)
+		if err != nil {
+			err = galasaErrors.NewGalasaError(galasaErrors.GALASA_ERROR_FAILED_TO_WRITE_FILE, targetFilePath, err.Error())
+		}
+	}
+	return newFile, err
+}
+
+// Retrieves an artifact for a given test run using its runId from the ecosystem API.
+func GetFileFromRestApi(runId string, artifactPath string, apiServerUrl string) (*os.File, error) {
+
+	var err error = nil
+	log.Printf("Downloading artifact '%s' from API server", artifactPath)
+
+	// A HTTP client which can communicate with the api server in an ecosystem.
+	restClient := api.InitialiseAPI(apiServerUrl)
+
+	fileDownloaded, httpResponse, err := restClient.ResultArchiveStoreAPIApi.
+		GetRasRunArtifactByPath(context.Background(), runId, artifactPath).
+		Execute()
+
+	if err != nil {
+		err = galasaErrors.NewGalasaError(galasaErrors.GALASA_ERROR_DOWNLOADING_ARTIFACT_FAILED, artifactPath, err.Error())
+	} else {
+		log.Printf("Downloaded artifact '%s' from API server OK", artifactPath)
+	}
+
+	defer httpResponse.Body.Close()
```

Comment:
why defer it ? Why not just close it ?

Reply:
Addressed in https://github.com/galasa-dev/cli/pull/92/commits/2f5f19fb10fe97c1ee2dd3523fb9327f83111106